### PR TITLE
Features/sntp

### PR DESCRIPTION
**Pull request by ricardochust**
SNTP working

you need to change the configuration on the .ioc lwip to enable sntp, which will generate a sntp.c in Middlewares/ThirdParty/lwip/src/apps/sntp/sntp.c

Then you will need to add a piece of code which configurates the sntp in Middlewares/ThirdParty/lwip/include/lwip/apps/sntp_opts.h just below the includes. This is the code:

```
static void set_time(uint32_t sec, uint32_t us);
static void set_rtc(uint16_t counter, uint8_t second, uint8_t minute, uint8_t hour, uint8_t day, uint8_t month, uint16_t year);
static u32_t get_rtc_s();
static u32_t get_rtc_us();

#define SNTP_STARTUP_DELAY 0
#define SNTP_SET_SYSTEM_TIME_US(sec,us) set_time(sec,us)
#define SUBSECONDS_PER_SECOND 32767
#define TRANSFORMATION_FACTOR (SUBSECONDS_PER_SECOND/999999.0)
#define SNTP_COMP_ROUNDTRIP 1
#define SNTP_UPDATE_DELAY 3600000

void set_time(uint32_t sec, uint32_t us){
	struct timeval tv;
	tv.tv_sec = sec;
	tv.tv_usec = us;
	time_t nowtime = sec;
	struct tm *nowtm = localtime(&nowtime);
	uint32_t subsecond = (uint32_t)(TRANSFORMATION_FACTOR * tv.tv_usec);
	set_rtc(subsecond, nowtm->tm_sec, nowtm->tm_min, nowtm->tm_hour, nowtm->tm_mday, 1+nowtm->tm_mon, 1900+(nowtm->tm_year));
}

#define SNTP_GET_SYSTEM_TIME(sec, us) do {(sec) = get_rtc_s(); (us) = get_rtc_us(); } while (0)

```

Also you will need to follow the Ethernet configuration (step 7 and below) as you will need to generate code so the sntp.c is created, which will remove the Ethernet configuration (and it is needed for sntp)

### CODE REVIEW

**Comment by ricardochust** [comment]
There is a bug where the year is not correctly set, but it is a bug on the rtc side and i will attend it on the rtc pull request

File: Inc/HALAL/Services/Communication/SNTP/SNTP.hpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
+/*
+ * SNTP.hpp
+ *
+ *  Created on: 21 feb. 2023
+ *      Author: Ricardo
+ */
+
+//https://community.st.com/s/question/0D53W00000j5I4aSAE/how-set-up-pretty-presice-and-detailed-rtc-on-stm32h743zi2
+
+//tontos usando sntp para configurar rtc
```
Comment (PabloGonzalezSegarra) [inline Inc/HALAL/Services/Communication/SNTP/SNTP.hpp:10]: Quita esto

File: Inc/HALAL/Services/Communication/SNTP/SNTP.hpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+#pragma once
+
+#include "C++Utilities/CppUtils.hpp"
+
+class SNTP{
+public:
+	static void sntp_update(uint8_t address_head, uint8_t address_second, uint8_t address_third, uint8_t address_last);
```
Comment (PabloGonzalezSegarra) [inline Inc/HALAL/Services/Communication/SNTP/SNTP.hpp:18]: Aqui seria mejor pasar un string como hace stefan en la clase ethernet o en el STLIB::start

File: Src/HALAL/Services/Communication/SNTP/SNTP.cpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+void SNTP::sntp_update(uint8_t address_head, uint8_t address_second, uint8_t address_third, uint8_t address_last){
+	sntp_setoperatingmode(SNTP_OPMODE_POLL);
+	ip4_addr_t* address;
+	IP_ADDR4(address,address_head,address_second,address_third,address_last);
+	sntp_setserver(0,address);
+	sntp_init();
+}
+
+extern "C" void set_rtc(uint16_t counter, uint8_t second, uint8_t minute, uint8_t hour, uint8_t day, uint8_t month, uint16_t year){
```
Comment (PabloGonzalezSegarra) [inline Src/HALAL/Services/Communication/SNTP/SNTP.cpp:26]: porque el extern?

**Review by PabloGonzalezSegarra — CHANGES_REQUESTED** [review]

File: Src/HALAL/Services/Communication/SNTP/SNTP.cpp (hunk 3 above)
Comment (ricardochust) [reply Src/HALAL/Services/Communication/SNTP/SNTP.cpp:26]: el extern es porque se usa dentro de sntp_opts.h, que esta en C. Le hace falta acceder al rtc para cambiar la fecha y esa funcion hace de puente para ello.

File: Inc/HALAL/Services/Communication/SNTP/SNTP.hpp (hunk 1 above)
Comment (ricardochust) [reply Inc/HALAL/Services/Communication/SNTP/SNTP.hpp:10]: ups

File: Inc/HALAL/Services/Communication/SNTP/SNTP.hpp (hunk 2 above)
Comment (ricardochust) [reply Inc/HALAL/Services/Communication/SNTP/SNTP.hpp:14]: done

File: Src/HALAL/Services/Communication/SNTP/SNTP.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+void SNTP::sntp_update(uint8_t address_head, uint8_t address_second, uint8_t address_third, uint8_t address_last){
+	sntp_setoperatingmode(SNTP_OPMODE_POLL);
+	ip4_addr_t* address;
+	IP_ADDR4(address,address_head,address_second,address_third,address_last);
+	sntp_setserver(0,address);
+	sntp_init();
+}
+
+void SNTP::sntp_update(string ip) {
```
Comment (StefanCostea) [inline Src/HALAL/Services/Communication/SNTP/SNTP.cpp:26]: esto es un start no un update

**Review by StefanCostea — APPROVED** [review]

**Review by PabloGonzalezSegarra — APPROVED** [review]

**Pull request merged**